R/init.R

```diff
@@ -122,52 +119,74 @@ init <- function(data,
-  checkmate::assert_multi_class(modules, c("teal_module", "list", "teal_modules"))
   checkmate::assert(
-    checkmate::check_class(filter, "teal_slices"),
-    checkmate::check_list(filter, names = "named")
+    .var.name = "data",
+    checkmate::check_multi_class(data, c("teal_data", "teal_data_module")),
+    checkmate::check_list(data, names = "named")
   )
-  checkmate::assert_multi_class(title, c("shiny.tag", "character"))
-  checkmate::assert_multi_class(header, c("shiny.tag", "character"))
-  checkmate::assert_multi_class(footer, c("shiny.tag", "character"))
-  checkmate::assert_character(id, max.len = 1, any.missing = FALSE)
-
-  teal.logger::log_system_info()
-
-  if (is.character(title)) {
-    title <- build_app_title(title)
-  } else {
-    validate_app_title_tag(title)
+  if (is.list(data) && !inherits(data, "teal_data_module")) {
+    data <- do.call(teal.data::teal_data, data)
   }
 
+  ## `modules`
+  checkmate::assert(
+    checkmate::check_multi_class(modules, c("teal_modules", "teal_module")),
+    checkmate::check_list(modules, min.len = 1, any.missing = FALSE, types = c("teal_module", "teal_modules"))
+  )
   if (inherits(modules, "teal_module")) {
     modules <- list(modules)
   }
-  if (inherits(modules, "list")) {
+  if (checkmate::test_list(modules, min.len = 1, any.missing = FALSE, types = c("teal_module", "teal_modules"))) {
     modules <- do.call(teal::modules, modules)
   }
 
+  ## `filter`
+  checkmate::assert(
+    .var.name = "filter",
+    checkmate::check_class(filter, "teal_slices"),
+    checkmate::check_list(filter, names = "named")
+  )
+
+  ## all other arguments
+  checkmate::assert(
+    .var.name = "title",
+    checkmate::check_string(title),
+    checkmate::check_multi_class(title, c("shiny.tag", "shiny.tag.list", "html"))
+  )
+  checkmate::assert(
+    .var.name = "header",
+    checkmate::check_string(header),
+    checkmate::check_multi_class(header, c("shiny.tag", "shiny.tag.list", "html"))
+  )
+  checkmate::assert(
+    .var.name = "footer",
+    checkmate::check_string(footer),
+    checkmate::check_multi_class(footer, c("shiny.tag", "shiny.tag.list", "html"))
+  )
+  checkmate::assert_character(id, max.len = 1, any.missing = FALSE)
+
+  # log
+  teal.logger::log_system_info()
+
+  # argument transformations
+  ## `modules` - landing module
   landing <- extract_module(modules, "teal_module_landing")
-  if (length(landing) > 1L) stop("Only one `landing_popup_module` can be used.")
-  modules <- drop_module(modules, "teal_module_landing")
-
-  # Calculate app id that will be used to stamp filter state snapshots.
-  # App id is a hash of the app's data and modules.
-  # See "transferring snapshots" section in ?snapshot.
-  hashables <- mget(c("data", "modules"))
-  hashables$data <- if (inherits(hashables$data, "teal_data")) {
-    as.list(hashables$data@env)
-  } else if (inherits(data, "teal_data_module")) {
-    body(data$server)
+  landing_module <- NULL
```

Comment:
I still don't like this. You are creating another variable just because you want to test for `NULL` and not `length==0`. 
You are also uncoupling executing the landing module from validating it (i.e. validating that there is only one). This may be down to preference because you _are_ coupling it to the extraction. 
Dropping modules under a condition - I don't know. If we ever add a module class that requires special handling, it will taken right back out.

I am not going to press it further but let it be noted.

Reply:
This is not only for replacing length check with NULL. This is rather an indirect consequence. My point was to keep the logic in one place close together. Namely: (i) landing module existence condition as well as (ii) extraction of the landing module used from all the landing modules we identified much closer to the landing modules extraction call. Currently (before this PR) we have both of these down below in `srv_teal_with_splash()` function call with a lot of other unrelated logic in between.
I really believe that this encapsulates the logic much better. This will make reading the code as well as potential future refactor much easier.